src/Raven.Server/Documents/ResourceCache.cs

```diff
@@ -86,44 +101,58 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
 
         public bool TryRemove(StringSegment resourceName, Task<TResource> resourceTask)
         {
-            if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
-                return false;
-
-            _resourceDetails.TryRemove(resourceTask, out _);
-
             lock (this)
             {
-                RemoveCaseSensitive(resourceName, resourceTask);
+                if (TryRemoveInternal(resourceName, resourceTask) == false) 
```

Comment:
Why this is under the lock now?

Reply:
There is a big change in how it used to work and how it works now. To avoid race conditions modifications now happen exclusively under critical section. There is no longer concurrency allowed in modification. In fact, I didn't change the concurrent dictionary for plain dictionary with the only purpose of being as close to the original as possible to avoid introducing too extensive changes.

Reply:
The reason why `.TryRemoveInternal()` is because construction of frozen dictionaries must be done at the end of any modification procedure before leaving the lock. So actual modification operations must not create new dictionaries.

Reply:
1. So basically all modifications are done under the lock. That is this big change that you're talkin about, right?
2. I don't follow the need of introducing `TryRemoveInternal()`

Reply:
It is a big change yes but it is not externally observable given the already racy semantics of the current code. Since we update dictionaries one at a time a reader may observe an unstable view (not a big deal for the usage pattern). 

Every remove and add actually now involves two stages. At stage 1 we do what we used to do in the method (now isolated in the internal one) and at stage 2 we create and update the references that are externally observable